lib/protobuf/field/field_hash.rb

```diff
@@ -0,0 +1,97 @@
+module Protobuf
+  module Field
+    class FieldHash < Hash
+
+      ##
+      # Attributes
+      #
+
+      attr_reader :field, :key_field, :value_field
+
+      ##
+      # Constructor
+      #
+
+      def initialize(field)
+        @field = field
+        @key_field = field.type_class.get_field(:key)
+        @value_field = field.type_class.get_field(:value)
+      end
+
+      ##
+      # Public Instance Methods
+      #
+
+      def []=(key, val)
+        super(normalize_key(key), normalize_val(val)) unless val.nil?
```

Comment:
Yeah, I wasn't really sure what to do. Maybe raise a type error if trying to put a nil in? For the case you pointed out in Slack (auto-adding nils when you random access an index past the tail), we might need to do something more in `FieldArray`, like bounds check before calling super?